datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java

```diff
@@ -0,0 +1,122 @@
+    String datastreamGroupName = datastreamGroup.getName();
+    Map<String, ClusterThroughputInfo> partitionThroughputInfo;
+
+    // Attempting to retrieve partition throughput info with a fallback mechanism to StickyPartitionAssignmentStrategy
+    try {
+      partitionThroughputInfo = fetchPartitionThroughputInfo();
+    } catch (RetriesExhaustedException ex) {
+      LOG.warn("Attempts to fetch partition throughput timed out. Falling back to regular partition assignment strategy");
+      return super.assignPartitions(currentAssignment, datastreamPartitions);
+    }
+
+    LOG.info("Old partition assignment info, assignment: {}", currentAssignment);
+    Validate.isTrue(currentAssignment.size() > 0,
+        "Zero tasks assigned. Retry leader partition assignment.");
+
+    // Calculating task count and assigned partitions
+    int taskCount = getTaskCountForDatastreamGroup(currentAssignment, datastreamGroupName);
+    List<String> assignedPartitions = getAssignedPartitionsForDatastreamGroup(currentAssignment, datastreamGroupName);
+
+    // Elastic task count validation
+    if (getEnableElasticTaskAssignment(datastreamGroup)) {
+      if (assignedPartitions.isEmpty()) {
+        performElasticTaskCountValidation(datastreamPartitions, taskCount);
+      }
+      updateOrRegisterElasticTaskAssignmentMetrics(datastreamPartitions, taskCount);
+    }
+
+    // Calculating unassigned partitions
+    List<String> unassignedPartitions = new ArrayList<>(datastreamPartitions.getPartitions());
+    unassignedPartitions.removeAll(assignedPartitions);
+
+    // Getting a task count estimate based on throughput
+    String clusterName = _sourceClusterResolver.getSourceCluster(datastreamPartitions.getDatastreamGroup());
+    LoadBasedTaskCountEstimator taskCountEstimator = new LoadBasedTaskCountEstimator();
+    ClusterThroughputInfo clusterThroughputInfo = partitionThroughputInfo.get(clusterName);
+    int throughputTaskCountEstimate = taskCountEstimator.getTaskCount(clusterThroughputInfo, assignedPartitions,
+        unassignedPartitions);
+
+    // Picking the maximum task count between elastic task count and throughput based task count estimate
+    int maxTaskCount = Math.max(taskCount, throughputTaskCountEstimate);
+
+    // Doing assignment
+    LoadBasedPartitionAssigner partitionAssigner = new LoadBasedPartitionAssigner();
+    return partitionAssigner.assignPartitions(clusterThroughputInfo, assignedPartitions, unassignedPartitions,
+        maxTaskCount);
+  }
+
+  private Map<String, ClusterThroughputInfo> fetchPartitionThroughputInfo() {
+    return PollUtils.poll(() -> {
+      try {
+        return _throughputProvider.getThroughputInfo();
+      } catch (Exception ex) {
+        LOG.warn("Failed to fetch partition throughput info.");
+        return null;
+      }
+    }, Objects::nonNull, THROUGHPUT_INFO_FETCH_RETRY_PERIOD_MS_DEFAULT, THROUGHPUT_INFO_FETCH_TIMEOUT_MS_DEFAULT)
```

Comment:
If there are capacity issues to serve this request, would it help to have exponential retries ?

Reply:
Our poll utilities don't support exponential timeout at the moment. But I think it will be useful in general to have that, I'll create a separate PR for that.